crates/types/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+//! This crate contains the raw base types (without implementation) for the World ID Protocol.
+//!
+//! It implements basic primitives such as field elements, proofs, the format of requests and responses, etc.
+//!
+//! Importantly, this crate keeps dependencies to a minimum and does not implement any logic beyond serialization and deserialization.
+#![deny(
+    clippy::all,
+    clippy::pedantic,
+    clippy::nursery,
+    missing_docs,
+    dead_code
+)]
+
+use std::{
+    fmt,
+    io::{Cursor, Read, Write},
+    str::FromStr,
+};
+
+use ark_ff::PrimeField;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ruint::aliases::U256;
+use serde::{de::Error as _, ser::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+
+/// Module containing the quintessential proof type.
+pub mod proof;
+pub use proof::WorldIdProof;
+
+/// Module containing types specifically related to relying parties.
+pub mod rp;
+
+/// The base field (curve field `Fq`) over which the elliptic curve is
+/// defined for the curve that is used to sign credentials in the World ID Protocol.
+///
+/// The World ID Protocol currently uses the `BabyJubJub` curve.
+pub type BaseField = ark_babyjubjub::Fq;
+
+/// Represents a Merkle root hash for any of the trees in the World ID Protocol.
+///
+/// The inner type is a base field element from `BabyJubJub` for convenience instead of a scalar field element on BN254.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct FieldElement(BaseField);
```

Comment:
I find it a bit confusing to have this wrapper for a wrapper. Maybe let's make `BaseField` already a custom type and do something like this?

```rust
pub struct BaseField(pub ark_babyjubjub::Fq);

impl Deref for BaseField {
    type Target = ark_babyjubjub::Fq;
    fn deref(&self) -> &Self::Target { &self.0 }
}
impl DerefMut for BaseField {
    fn deref_mut(&mut self) -> &mut Self::Target { &mut self.0 }
}
```